### ISSUE TITLE
Invoke build `-if` parameter scriptblock evaluation

### ISSUE THREAD

**Issue by kapilmb** [nightroman/Invoke-Build#51]
Consider the following build script
```powershell
# c:\tmp\test-invokebuild.build.ps1
task Build1 -Before "Build2" { $script:a = "a" }
task Build2 -if {$script:a} { }
```

```
C:\tmp
PS> Invoke-Build build2
Build build2 C:\tmp\test-invokebuild.build.ps1
Task /Build2 skipped.
Build succeeded. 0 tasks, 0 errors, 0 warnings 00:00:00.0300201

C:\tmp
PS> Invoke-Build build1,build2
Build build1, build2 C:\tmp\test-invokebuild.build.
Task /Build1
Done /Build1 00:00:00.0130020
Done /Build2/Build1
Task /Build2
Done /Build2 00:00:00.0160129
Build succeeded. 2 tasks, 0 errors, 0 warnings 00:00:00.0620040
```

Looks like the conditional in task `Build2` is evaluated before invoking task `Build1` even though `Build2` depends on `Build1`. Is this by design?

I think the conditional in `Build2` should be evaluated after the invocation of `Build1`. Consider the following scenario:
```powershell
task buildA -before  BuildB,BuildC { $script:invokeBuildB = Test-Something }
task buildB -if {$script:invokeBuildB} -Before BuildD { ... }
task buildC  -if {!$script:invokeBuildB} -Before BuildD { ... }
task BuildD { ... }
```

I want the build script to take either path `A -> B -> D` or path `A -> C -> D` depending on what happens in `A`. I can add an `if` statement in `B` and `C` to achieve this but having the invokebuild engine do that for me would be nice.

**Comment by nightroman** [comment]
> Looks like the conditional in task Build2 is evaluated before invoking task
Build1 even though Build2 depends on Build1. Is this by design?

Yes, this is by design that a condition is evaluated first, that is before
referenced tasks.

The whole idea is that if a condition evaluates to false then the task and
*its references* are not invoked.

Let's take a look at the similar MSBuild example:

````xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Target Name="Test" Condition="Exists ('z.tmp')" DependsOnTargets="Target1">
        <Message Text="In Test"/>
    </Target>
    <Target Name="Target1">
      <Message Text="In Target1"/>
      <Copy SourceFiles="z.txt" DestinationFiles="z.tmp"/>
    </Target>
</Project>
````

Example explained:

- `Test` depends on `Target1`.
- `Target1` is supposed to create `z.tmp`.
- `Test` is invoked only if `z.tmp` exists.
- Deadlock...

That means MSBuild evaluates conditions before referenced targets.
And Invoke-Build evaluates conditions before referenced tasks.

**Comment by nightroman** [comment]
All in all, I think we can close the issue. But could you describe your real case? I am sure together we can find some suitable solution.

**Comment by nightroman** [comment]
I have just found some weird difference between MSBuild and Invoke-Build. But
even with this difference we can see that the condition is evaluated before
other tasks are invoked.

Example is similar to the above but instead of `DependsOnTargets="Target1"` on `Test`
we use `BeforeTargets="Test"` on `Target1`.

In Invoke-Build such a swap makes no difference (order of some tasks make change, though).

In MSBuild this swap makes `Target1` invoked. But the condition is still checked before
and as a result the target `Test` is actually skipped.

Given, we have `z.txt` and we do not have `z.tmp` in the script directory.
Invoke this MSBuild script:

<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Target Name="Test" Condition="Exists ('z.tmp')">
        <Message Text="In Test"/>
    </Target>
    <Target Name="Target1" BeforeTargets="Test">
      <Message Text="In Target1"/>
      <Copy SourceFiles="z.txt" DestinationFiles="z.tmp"/>
    </Target>
</Project>

Log

- Target "Test" skipped, due to false condition.
- We can see `In Target1`, it is invoked!
- We cannot see `In Test`, it is skipped, even though at this point its condition would get true.

In other words, in Invoke-Build (unlike in MSBuild) these two scripts are equivalent:

    # script1
    task Build1 {...}
    task Build2 -If {...} Build1, {...}

    # script2
    task Build1 -Before Build2 {...}
    task Build2 -If {...} {...}

This is by design and I think it is correct to keep it this way.
I will mention the difference at [Comparison-with-MSBuild](https://github.com/nightroman/Invoke-Build/wiki/Comparison-with-MSBuild).

But even with this difference, both MSBuild and Invoke-Build evaluate
conditions before other involved tasks/targets. The examples show this.

**Comment by nightroman** [comment]
To your example.

1) Here it is (not working as we want):

````powershell
    task buildA -before  BuildB,BuildC { $script:invokeBuildB = $true } # or $false
    task buildB -if {$script:invokeBuildB} -Before BuildD { }
    task buildC -if {!$script:invokeBuildB} -Before BuildD { }
    task BuildD { }
````

2) It is identical to (still not working as we want)

````powershell
    task buildA { $script:invokeBuildB = $true } # or $false
    task buildB -if {$script:invokeBuildB} buildA, { }
    task buildC -if {!$script:invokeBuildB} buildA, { }
    task buildD buildB, buildC, { }
````

3) This works as we want

````powershell
    task buildA { $script:invokeBuildB = $true } # or $false
    task buildB buildA, {
        if ($script:invokeBuildB) { } else { }
    }
    task buildD buildB, { }
````

Is 3) is not suitable, please tell why and we will find something else.

In other words, as far as one and only one of the tasks B and C is invoked after A, then logically we do not have two tasks. We have one task "process results of A". That is what we do in 3).

**Comment by kapilmb** [comment]
@nightroman Thanks for the detailed explanation! I understand the rationale behind the behavior now. I was playing with invokebuild and was wondering about the situation for which I created the issue. As of now I do not have a concrete situation where I would need this. Thanks again!

**Comment by nightroman** [comment]
@kapilmb No problem! It was useful to look at this again. In particular, a not yet known difference between MSBuild and IB was found. I documented it in [Comparison-with-MSBuild](https://github.com/nightroman/Invoke-Build/wiki/Comparison-with-MSBuild). I do not know what scenarios are covered by MSBuild and how useful they are. The IB's way looks practical and convenient.
